### PR TITLE
feat(judicial-system): Archive Old Cases

### PR DESCRIPTION
**Pull request by gudjong**
# Archive Old Cases

https://app.asana.com/0/1199153462262248/1201714251473992/f

## What

- Archives old cases 90 days after ruling or end of custody.
- Archives deleted cases.
- Encrypts selected database properties in a json object when archiving a case.
- Stores the encrypted json object in a new table `case_archive`.
- Clears the encrypted database properties.
- Addresses some linting and naming issues.

## Why

Security reasons - sensitive information should not be accessible longer than necessary.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

**Review by pshomov — APPROVED** [review]
LGTM

File: apps/judicial-system/backend/src/app/modules/case/filters/case.filters.ts [hunk 1]
```diff
@@ -151,6 +153,70 @@ function isHightenedSecurityCaseHiddenFromUser(
...
+            decision: [
+              CaseDecision.ACCEPTING,
+              CaseDecision.ACCEPTING_PARTIALLY,
+            ],
+          },
+        ],
+      }
+}
+
+export const oldFilter = {
```
Comment (Arnarkari93) [inline apps/judicial-system/backend/src/app/modules/case/filters/case.filters.ts:182]: Nice

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts [hunk 2]
```diff
@@ -863,4 +863,14 @@ export class Case extends Model<Case> {
   })
   @ApiProperty()
   caseResentExplanation?: string
+
+  /**********
+   * Indicates whether the case has been archived - optional
+   **********/
+  @Column({
+    type: DataType.BOOLEAN,
+    allowNull: true,
```
Comment (Arnarkari93) [inline apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:872]: Should we not allow a null value here?
```suggestion
    allowNull: false,
```

File: apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+          [{ model: Defendant, as: 'defendants' }, 'created', 'ASC'],
+          [{ model: CaseFile, as: 'caseFiles' }, 'created', 'ASC'],
+        ],
+        where: {
+          isArchived: false,
+          [Op.or]: [
+            { state: CaseState.DELETED },
+            {
+              [Op.or]: [
+                {
```
Comment (Arnarkari93) [inline apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts:123]: I don't really see the value in checking if `caseModel.fineOne` has been called with the right parameters. Can we test the return value? 
If we are not able to test the return value, can we reuse the `oldFilter` here?

File: apps/judicial-system/backend/src/app/modules/file/file.service.ts [hunk 4]
```diff
@@ -233,4 +235,33 @@ export class FileService {
...
+    if (numberOfAffectedRows > 1) {
+      // Tolerate failure, but log error
+      this.logger.error(
+        `Unexpected number of rows (${numberOfAffectedRows}) affected when updating file ${fileId} of case ${caseId}`,
+      )
+    } else if (numberOfAffectedRows < 1) {
+      throw new InternalServerErrorException(
+        `Could not update file ${fileId} of case ${caseId}`,
+      )
+    }
```
Comment (Arnarkari93) [inline apps/judicial-system/backend/src/app/modules/file/file.service.ts:265]: Do we want to return `{ success }` here like the other methods in the service?

File: apps/judicial-system/scheduler/src/app/app.config.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+export const appModuleConfig = defineConfig({
+  name: 'AppModule',
+  load: (env) => ({
+    backendAccessToken: env.required(
+      'BACKEND_ACCESS_TOKEN',
+      'secret-backend-api-token',
+    ),
+    backendUrl: env.required('BACKEND_URL', 'http://localhost:3344'),
+    timeToLiveMinutes: env.requiredJSON('TIME_TO_Live_MINUTES', 1),
```
Comment (Arnarkari93) [inline apps/judicial-system/scheduler/src/app/app.config.ts:11]: Should this be all caps?
```suggestion
    timeToLiveMinutes: env.requiredJSON('TIME_TO_LIVE_MINUTES', 1),
```

**Review by Arnarkari93 — COMMENTED** [review]
Great job! I just have a couple of questions.

File: apps/judicial-system/backend/src/app/modules/case/models/case.model.ts (hunk 2 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/models/case.model.ts:872]: Agreed, that is in line with the database migration.

File: apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts (hunk 3 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts:123]: I am mocking the return value and that is the reason I test the call parameters. We have no way of testing the return value unless we call a running database and that involves a pretty hefty test setup cost.

Basically, I am saying: If our code makes the correct call, then I trust the database to return the expected value.

Using oldFilter in the test is a good idea.

File: apps/judicial-system/backend/src/app/modules/file/file.service.ts (hunk 4 above)
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/file/file.service.ts:267]: Thank you for pointing out the discrepancy. Actually, there are other methods that do not return `{ success }`, but all the other methods return something. The reason I did not do that here is that this method's return value will not be passed to the client and is not used by the caller (in this case the `archive` method in the `CaseService`. However, to stick to the patterned used elsewhere on the server, this method should return the updated case file. I have made that change.

File: apps/judicial-system/backend/src/app/modules/case/case.controller.ts [hunk 6]
```diff
@@ -526,4 +522,16 @@ export class CaseController {
...
   }
+
+  @UseGuards(TokenGuard)
+  @Post('internal/cases/archive')
+  @ApiOkResponse({
+    type: ArchiveResponse,
+    description: 'Archives a single case if any case is ready to be archived',
+  })
+  archive(): Promise<ArchiveResponse> {
+    this.logger.debug('Archiving a case')
```
[the remark this answers is not in the document]
Comment (Arnarkari93) [reply apps/judicial-system/backend/src/app/modules/case/case.controller.ts:543]: Sounds good!

File: apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts (hunk 3 above)
Comment (Arnarkari93) [reply apps/judicial-system/backend/src/app/modules/case/test/archive.spec.ts]: I see, then this makes sense.

**Review by Arnarkari93 — APPROVED** [review]
🕵️‍♀️ LGTM

**Pull request merged**